scene/main/viewport.cpp

```diff
@@ -58,6 +58,10 @@
 #include "servers/audio_server.h"
 #include "servers/rendering/rendering_server_globals.h"
 
+#ifdef MODULE_GRIDMAP_ENABLED
+#include "modules/gridmap/grid_map.h"
+#endif
```

Comment:
I don't think having GridMap creep its way into something as core as Viewport will be acceptable.

I'm not sure how this should be done but this is probably not the way.

Reply:
Could a `unhandled_input_picking` signal from the viewport work? That way GridMap can connect to it and receive the information needed. If yes, should the signal be exposed to GDScript?

Reply:
That might work, though I'm surprised that there wouldn't be an existing system that can be suitable for GridMap. It's not the only node / editor that handles picking and if other gizmos can do it, then GridMap should be able to do it too (just talking in principle - I'm not actually familiar with that code and might be wrong).

CC @JFonS @RandomShaper

Reply:
Well, that can certainly be done with the normal `unhandled_input`. But since the viewport is already checking for collisions I think that can be used and avoid repeating code. Also, the `process_picking` method (the one that takes care of collisions) is called inside `physics_process` of the SceneTree. If we want the input to also be processed in GridMap's `physics_process`, the inputs should be collected first and then processed. To me, that's a bit too much for something that is already implemented and can be reused.

Reply:
I would be interested in having a generic way to connect to this, because this issue is not specific to GridMap: https://github.com/godotengine/godot-proposals/issues/3818#issuecomment-1902687054